File: 백준/Gold/2294. 동전 2/동전 2.cc
#include <iostream>
#include <algorithm>
#include <vector>
#define MAX 10003
#define CMAX 103
#define INF 987654321
using namespace std;
int dp[MAX];
vector<int> coin;
int n, k;
void input() {
	cin >> n >> k;
	int t;
	for (int i = 0; i < n; i++) {
		cin >> t;
		coin.push_back(t);
	}
	fill(dp + 1, dp + MAX, INF);
	sort(coin.begin(), coin.end());
}
int solve() {
	
	for (int i = 1; i <= k; i++) {
		for (int j = 0, index = 0; j < n; j++) {

			if (coin[j] <= i) {
				index = i - coin[j];
				dp[i] = min(dp[i], dp[index] + 1);
			} 
		}
		
	}
	return dp[k] == INF ? -1 : dp[k];
}
int main() {
	cin.tie(0); ios::sync_with_stdio(0);
	input();
	cout << solve();
}

File: 백준/Gold/2294. 동전 2/동전 2.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static StringBuilder sb = new StringBuilder();
	static int n,k;
	static final int INF = 987654321;
	static int dp[],coin[];
	static void input() throws Exception {
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		k = Integer.parseInt(st.nextToken());
		dp = new int[k+1];
		coin = new int[n];
		for (int i = 0; i < n; i++) {
			coin[i] = Integer.parseInt(br.readLine());
		}
		for (int i = 1; i <= k; i++) {
			dp[i] = INF;
		}
		Arrays.sort(coin);
		
	}
	static int solve() throws Exception {
		int res=0;
		
		for (int i = 1; i <= k; i++) { // 원
			for (int j = 0, index = 0; j < n; j++) { // 코인
				if(coin[j] <= i) { // 인덱스에 넣을 수 있는 경우
					index = i - coin[j];
					dp[i] = Math.min(dp[i], dp[index]+1);
				}
			}
		}
		if(dp[k] == INF) res = -1;
		else res = dp[k];
		return res;
		
	}
	public static void main(String[] args) throws Exception {
		input();
		System.out.println(solve());
	}
}